hail/src/main/scala/is/hail/expr/ir/Env.scala

```diff
@@ -220,19 +251,28 @@ case class BindingEnv[V](
     }
   }
 
-  def subtract(newBindings: BindingEnv[_]): BindingEnv[V] = {
-    if (agg.isDefined != newBindings.agg.isDefined || scan.isDefined != newBindings.scan.isDefined)
+  def subtract(
+    eval: Seq[(String, V)],
+    agg: Option[Seq[(String, V)]],
+    scan: Option[Seq[(String, V)]],
+    relational: Seq[(String, V)],
+  ): BindingEnv[V] = {
+    if (this.agg.isDefined != agg.isDefined || this.scan.isDefined != scan.isDefined)
       throw new RuntimeException(s"found inconsistent agg or scan environments:" +
-        s"\n  left: ${agg.isDefined}, ${scan.isDefined}" +
-        s"\n  right: ${newBindings.agg.isDefined}, ${newBindings.scan.isDefined}")
-    if (allEmpty || newBindings.allEmpty)
+        s"\n  left: ${this.agg.isDefined}, ${this.scan.isDefined}" +
+        s"\n  right: ${agg.isDefined}, ${scan.isDefined}")
```

Comment:
This is exactly the message that would be a million times better than `None.get` above - maybe you can pull this out into a function and use that in your other suberact implementation?

Reply:
I think I found a good partial solution. I put good error messages in `BindingEnv.modifyWithoutNewBindings`, which most things now go through.